src/tomli/_parser.py

```diff
@@ -61,7 +62,30 @@ class TOMLDecodeError(ValueError):
     colno: The column corresponding to pos
     """
 
-    def __init__(self, msg: str, doc: str, pos: Pos):
+    def __init__(
+        self,
+        *args: Any,
+        msg: str | None = None,
+        doc: str | None = None,
+        pos: Pos | None = None,
+    ):
+        if args:
+            warnings.warn(
+                "Positional arguments are deprecated. "
+                "Please set 'msg', 'doc' and 'pos' keyword arguments.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+            ValueError.__init__(self, *args)
+            return
```

Comment:
I would set the new instance variables to None here.

Reply:
What problems could this cause? If we don't set the variables, we can promise non-None types for the attributes. And since the attributes never were there when using positional args in `__init__` this shouldn't break anything.

Reply:
No, this is worse, because now you need to use `getattr()`, because the attributes might not even exist:

```python
msg = getattr(exc, "msg")
doc = getattr(exc, "doc")
pos = getattr(exc, "pos")
lineno = getattr(exc, "lineno")
colno = getattr(exc, "colno")
print(f"{msg}: line {lineno} column {colno} (char {pos})")
```

Reply:
The attributes will exist 100% of the time when raised by the library and caught by a consumer of it, though. You only need getattr in the obscure case someone created an instance themselves.

Reply:
A linter will complain about this, and we also know that the attributes can't be None when raised by the library.

Reply:
Hmm, which linter will complain? Mypy and flake8 seem to accept this just fine, they run in CI.

I don't feel strongly about this, I'm happy to add the Nones if cpython codeowners prefer that. If the None case stays deprecated, perhaps I can even persuade typeshed to conveniently lie about the types never being None :)

Reply:
Hmm, they don't.

Reply:
I'm happy to know that, and sorry, my bad :smile_cat: